doc/api/process.md

```diff
@@ -1881,6 +1881,271 @@ a code.
 Specifying a code to [`process.exit(code)`][`process.exit()`] will override any
 previous setting of `process.exitCode`.
 
+## `process.finalization.register(ref, callback)`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1 - Experimental
+
+* `ref` {Object | Function} The reference to the resource that is being tracked.
+* `callback` {Function} The callback function to be called when the resource
+  is finalized.
+  * `ref` {Object | Function} The reference to the resource that is being tracked.
+  * `event` {string} The event that triggered the finalization. Defaults to 'exit'.
+
+This function registers a callback to be called when the process emits the `exit` event.
+
+If the object is garbage collected before `exit` event is emitted, the callback will be
+removed from the finalization registry, and it will not be called on process exit.
+
+Inside the callback you can do whatever you want with the object,
+including the release of resources allocated by that resource.
+
+The idea of ​​this function is to help you free up resources when the process exit,
+but also let the object be garbage collected if it is no longer being used.
+
+Ex: you can register a stream to be released when the process exit, if the stream is
+garbage collected before the process exit, we no longer need to release the stream, so in this case
+we just remove the callback from the finalization registry.
+
+Caveats: This function relies on the `FinalizationRegistry` API, which the
+authors of this API said to "avoid where possible" because the nature of
+how the garbage collector works differently between engines, there's no
+specific behavior guarantee by the specification. In their words:
+"Developers shouldn't rely on cleanup callbacks for essential program
+logic. Cleanup callbacks may be useful for reducing memory usage across
+the course of a program, but are unlikely to be useful otherwise". You
+can read more about the caveats at [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry#avoid_where_possible)
+or in the [proposal WeakRefs](https://github.com/tc39/proposal-weakrefs?tab=readme-ov-file#a-note-of-caution).
+
+```cjs
+const { finalization, stdout } = require('node:process');
+
+function setup() {
+  // This object can be safely garbage collected,
+  // and the resulting shutdown function will not be called.
+  // There are no leaks.
+  const myDisposableObject = {
+    dispose() {
+      // Free your resources synchronously
+      stdout.write('disposed.\n');
+    },
+  };
+
+  // Please make sure that the function passed to finalization.register()
+  // does not create a closure around unnecessary objects.
+  function onFinalize(obj, event) {
+    // You can do whatever you want with the object
+    obj.dispose();
+  }
```

Comment:
Can you give me a more detailed explanation about why this is not correct?

Because I think this could be a very common way of using this API:

```js
finalization.register(obj, (ref) => {
  ref.dispose();
});
```

Reply:
If you keep the `onFinalize` function inside the `setup` function, it will close upon the `myDisposableObject` and therefore it will _never_ be collected. You need to move `onFinalize` outside of the `setup` function.

Reply:
Are you sure about that? There's no reason for the engine to close on `myDisposableObject`. It's not referenced by the function.

Reply:
It doesn't work that way. The function references the previous closure, which includes `myDisposableObject`.

Reply:
I will update the documentation to something like:

```md
The code above uses the following assumptions:
- arrow function is avoided
- regular function is recommended to be on global context (root)

Regular functions *could* reference the context where the `obj` lives, making the `obj` not garbage collectible. 

About arrow functions, they will hold the previous context, so if you write a code like this:

```js
class Test {
  constructor() {
     finalization.register(this, () => this.dispose())
   }
   dispose() {}
}
\```

This object is very unlikely (not impossible) to be garbage collectible, but at least the `dispose` will be called when the `process.exit`.

Again, everything we talked about GC is assumptions that could be changed in the future, so be aware and avoid relying critical things on this feature.
```